src/data/quaternion.lean

```diff
@@ -0,0 +1,318 @@
+/-
+Copyright (c) 2020 Yury Kudryashov. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yury Kudryashov
+-/
+
+import tactic.ring_exp ring_theory.algebra algebra.opposites algebra.commute data.equiv.ring
+
+/-!
+# Quaternions
+
+In this file we define quaternions `ℍ[R]` over a commutative ring `R`, and define some
+algebraic structures on `ℍ[R]`.
+
+## Main definitions
+
+* `quaternion R`, `ℍ[R]` : the space of quaternions;
+* `quaternion.norm_sq` : square of the norm of a quaternion;
+* `quaternion.conj` : conjugate of a quaternion;
+
+We also define the following algebraic structures on `ℍ[R]`:
+
+* `ring ℍ[R]` and `algebra R ℍ[R]` : for any commutative ring `R`;
+* `domain ℍ[R]` : for a linear ordered commutative ring `R`;
+* `division_algebra ℍ[R]` : for a linear ordered field `R`.
+
+## Notation
+
+* `ℍ[R]` : quaternions over `R`.
+
+## Implementation notes
+
+We define quaternions over any ring `R`, not just `ℝ` to be able to deal with, e.g., integer
+or rational quaternions without using real numbers. In particular, all definitions in this file
+are computable.
+
+## Tags
+
+quaternion
+-/
+
+/-- Space of quaternions over a type. Implemented as a structure with four fields:
+`re`, `im_i`, `im_j`, and `im_k`. -/
+@[ext] structure quaternion (R : Type*) :=
+(re : R) (im_i : R) (im_j : R) (im_k : R)
+
+notation `ℍ[` R `]` := quaternion R
+
+namespace quaternion
+
+variables {R : Type*} [comm_ring R] (r x y z : R) (a b c : ℍ[R])
+
+instance : has_coe_t R ℍ[R] := ⟨λ x, ⟨x, 0, 0, 0⟩⟩
+
+mk_simp_attribute quaternion_projs "Projections of operations on quaternions"
+
+@[simp, quaternion_projs] lemma coe_re : (x : ℍ[R]).re = x := rfl
+@[simp, quaternion_projs] lemma coe_im_i : (x : ℍ[R]).im_i = 0 := rfl
+@[simp, quaternion_projs] lemma coe_im_j : (x : ℍ[R]).im_j = 0 := rfl
+@[simp, quaternion_projs] lemma coe_im_k : (x : ℍ[R]).im_k = 0 := rfl
+
+lemma coe_injective : function.injective (coe : R → ℍ[R]) :=
+λ x y h, congr_arg re h
+
+@[simp] lemma coe_inj {x y : R} : (x : ℍ[R]) = y ↔ x = y := coe_injective.eq_iff
+
+@[simps short_name] instance : has_zero ℍ[R] := ⟨⟨0, 0, 0, 0⟩⟩
+
+@[simp, elim_cast] lemma coe_zero : ((0 : R) : ℍ[R]) = 0 := rfl
+
+instance : inhabited ℍ[R] := ⟨0⟩
+
+@[simps short_name] instance : has_one ℍ[R] := ⟨⟨1, 0, 0, 0⟩⟩
+
+@[simp, elim_cast] lemma coe_one : ((1 : R) : ℍ[R]) = 1 := rfl
+
+@[simps short_name] instance : has_add ℍ[R] := ⟨λ a b, ⟨a.1 + b.1, a.2 + b.2, a.3 + b.3, a.4 + b.4⟩⟩
+
+@[simps short_name] instance : has_neg ℍ[R] := ⟨λ a, ⟨-a.1, -a.2, -a.3, -a.4⟩⟩
+
+@[simps short_name] instance : has_mul ℍ[R] := ⟨λ a b,
```

Comment:
Over a field that is not real closed there are many non-isomorphic candidates for the quaternions. I wonder if we should take that into account in this PR. It is common to turn `i,j,k` into a sort of distinguished standard basis... but if you make a different choice, you would get a different quaternion algebra over the rationals.
See also https://en.wikipedia.org/wiki/Quaternion_algebra#Structure which shows that the multiplication depends on two parameters in general...

Reply:
I'm not an expert here. I decided to add quaternions just for fun, then decided to have a parameter `R` because I thought that some class instances are slow because `real` have too many instances (it didn't help; using `ring_exp` instead of `ring` helped).

So, what do you propose? Rename `quaternion` to `quaternion_algebra` and add two extra parameters?

Reply:
I think I would. And then reserve the `\H[R]` notation for the case where the parameters are `-1, -1`. I hope this isn't too much work.

Reply:
Done.